fuel-core/src/executor.rs

```diff
@@ -2449,6 +2567,281 @@ mod tests {
+        // modifies the state and the balance is successful, it should update roots.
+        let mut rng = StdRng::seed_from_u64(2322u64);
+
+        // Create a contract that modifies the state
+        let (create, contract_id) = create_contract(
+            vec![
+                // Sets the state STATE[0x1; 32] = value of `REG_PC`;
+                Opcode::SWW(0x1, 0x29, REG_PC),
+                Opcode::RET(1),
+            ]
+            .into_iter()
+            .collect::<Vec<u8>>(),
+            &mut rng,
+        );
+
+        let transfer_amount = 100 as Word;
+        let asset_id = AssetId::from([2; 32]);
+        let (script, data_offset) = script_with_data_offset!(
+            data_offset,
+            vec![
+                // Set register `0x10` to `Call`
+                Opcode::MOVI(0x10, (data_offset + AssetId::LEN as u32) as Immediate18),
+                // Set register `0x11` with offset to data that contains `asset_id`
+                Opcode::MOVI(0x11, data_offset),
+                // Set register `0x12` with `transfer_amount`
+                Opcode::MOVI(0x12, transfer_amount as u32),
+                Opcode::CALL(0x10, 0x12, 0x11, REG_CGAS),
+                Opcode::RET(REG_ONE),
+            ],
+            ConsensusParameters::DEFAULT.tx_offset()
+        );
+
+        let script_data: Vec<u8> = [
+            asset_id.as_ref(),
+            Call::new(contract_id, transfer_amount, data_offset as Word)
+                .to_bytes()
+                .as_ref(),
+        ]
+        .into_iter()
+        .flatten()
+        .copied()
+        .collect();
+
+        let modify_balance_and_state_tx = TxBuilder::new(2322)
+            .gas_limit(10000)
+            .coin_input(AssetId::zeroed(), 10000)
+            .start_script(script, script_data)
+            .contract_input(contract_id)
+            .coin_input(asset_id, transfer_amount)
+            .contract_output(&contract_id)
+            .build()
+            .transaction()
+            .clone();
+        let db = &mut Database::default();
+
+        let executor = Executor {
+            database: db.clone(),
+            config: Config {
+                utxo_validation: false,
+                ..Config::local_node()
+            },
+        };
+
+        let block = PartialFuelBlock {
+            header: PartialFuelBlockHeader {
+                consensus: FuelConsensusHeader {
+                    height: 1u64.into(),
+                    ..Default::default()
+                },
+                ..Default::default()
+            },
+            transactions: vec![create.into(), modify_balance_and_state_tx.into()],
+        };
+
+        let ExecutionResult {
+            block, tx_status, ..
+        } = executor.execute(ExecutionBlock::Production(block)).unwrap();
+
+        let empty_state = Bytes32::from(*empty_sum());
+        let executed_tx = block.transactions()[2].as_script().unwrap();
```

Comment:
we should also be testing:

1. that the tx in the db matches the tx in the block
2. that utxos in the db match the expected state, since those are separate from the output types on the tx

related: https://github.com/FuelLabs/fuel-core/pull/789/files#r1040374277